src/System.Management.Automation/engine/MshCommandRuntime.cs

```diff
@@ -3018,18 +3123,34 @@ internal ActionPreference WarningPreference
             {
                 // Setting CommonParameters.WarningAction has highest priority
                 if (IsWarningActionSet)
+                {
                     return _warningPreference;
+                }
 
                 if (Debug)
+                {
                     return ActionPreference.Inquire;
+                }
+
                 if (Verbose)
+                {
                     return ActionPreference.Continue;
+                }
+
                 // Debug:$false and Verbose:$false ignored
 
                 if (!_isWarningPreferenceCached)
                 {
-                    bool defaultUsed = false;
-                    _warningPreference = Context.GetEnumPreference<ActionPreference>(SpecialVariables.WarningPreferenceVarPath, _warningPreference, out defaultUsed);
+                    _warningPreference = Context.GetEnumPreference<ActionPreference>(SpecialVariables.WarningPreferenceVarPath, _warningPreference, out _);
+
+                    // If the host couldn't prompt for the warning action anyways, change it to 'Continue'.
+                    // This lets hosts still see debug output without having to implement the prompting logic.
+                    if ((CBhost.ExternalHost.UI == null) && (_warningPreference == ActionPreference.Continue))
+                    {
+                        _warningPreference = ActionPreference.Continue;
+                    }
+
+                    _isWarningPreferenceCached = true;
```

Comment:
FYI for reviewers, lines 3146-3153 were added because it seemed that they were missing (we were checking if the warning preference was cached or not, but never actually marking it as cached).

Reply:
After some more debugging and reading, I figured out how this works as well. Caching applies when you use `PSCmdlet.WriteVerbose`, `PSCmdlet.WriteWarning`, `PSCmdlet.WriteError`, etc., and it is supposed to honor the preference that was passed into the invocation of the command. For example, consider this code:

```powershell
# Make sure preferences are configured to show all messages
Get-Variable *Preference `
    | Where-Object Value -is [System.Management.Automation.ActionPreference] `
    | ForEach-Object {$_.Value = [System.Management.Automation.ActionPreference]::SilentlyContinue}

# Define a function that tests whether or not caching is working properly
function Test-InvocationCaching {
    [CmdletBinding()]
    param()
    $PSCmdlet.WriteError(
        [System.Management.Automation.ErrorRecord]::new(
            [System.Management.Automation.CommandNotFoundException]::new('Hello'),
            'FirstError',
            'InvalidOperation',
            $null)) # The current ErrorPreference value is cached here
    $ErrorActionPreference = 'Continue'
    $PSCmdlet.WriteError(
        [System.Management.Automation.ErrorRecord]::new(
            [System.Management.Automation.CommandNotFoundException]::new('World'),
            'SecondError',
            'InvalidOperation',
            $null)) # The cached ErrorPreference value should be used here
    $PSCmdlet.WriteWarning('Hello') # The current WarningPreference value is cached here
    $WarningPreference = 'Continue'
    $PSCmdlet.WriteWarning('World') # The cached WarningPreference value should be used here
    $PSCmdlet.WriteVerbose('Hello') # The current VerbosePreference value is cached here
    $VerbosePreference = 'Continue'
    $PSCmdlet.WriteVerbose('World') # The cached VerbosePreference value should be used here
    $PSCmdlet.WriteDebug('Hello') # The current DebugPreference value is cached here
    $DebugPreference = 'Continue'
    $PSCmdlet.WriteDebug('World') # The cached DebugPreference value should be used here
    $PSCmdlet.WriteInformation('Hello', $null) # The current InformationPreference value is cached here
    $InformationPreference = 'Continue'
    $PSCmdlet.WriteInformation('World', $null) # The cached InformationPreference value should be used here
}

# Now run the function
Test-InvocationCaching
```

If you run that script as is on any version of PowerShell that does not have this PR merged in, you will see one warning message and one verbose message output to the console. This occurs because the verbose preference and warning preference caching handling was missing, but it is properly set and used for error, debug, and information messages.

The fact that the warning and verbose messages show up in this case is technically incorrect, because the `PSCmdlet.WriteVerbose` and `PSCmdlet.WriteWarning` methods (like all other `PSCmdlet.Write*` methods) are supposed to cache the cmdlet invocation preference the first time they need to reference them and then use that cached value for the remainder of the invocation of the cmdlet. The fact that they don't do that today is a bug, and potentially a performance issue in a command that produces many warning (unlikely) and/or verbose (possible) messages.

Based on this analysis, I think I should keep the caching behavior that I have updated to work properly for Verbose and Warning, and add Pester tests to ensure that it works that way. Changing this won't break anything, but it may expose a bug in code that is doing weird things with `PSCmdlet` methods and preference variables; however, I think that's pretty unlikely.

@PaulHigin If you agree, I'll proceed with adding the Pester tests I talked about to ensure this continues to work as designed going forward.

Reply:
Ok, this sort of makes sense.  It seems like 'cache' is not the right term to use, and instead the value is set only once and needs to be retrieved and persist for the lifetime of the command (MshCommandRuntime).
Yes, please fix the Verbose,Warning cases and also add a comment explaining the need for this code logic.

Reply:
Will do.

Reply:
Ok, I added comments and renamed the fields named with cache to something more appropriate. This part is resolved. I'm just going to add a Pester test to ensure the variables are properly read only once each.

Reply:
Just realized I never circled back to say the test I was going to add is added. This is resolved.